src/Uno.Toolkit.UI/Controls/ZoomContentControl/ZoomContentControl.cs

```diff
@@ -0,0 +1,482 @@
+#endif
+			e.KeyModifiers.HasFlag(Windows.System.VirtualKeyModifiers.Control) &&
+			IsZoomAllowed)
+		{
+			e.Handled = true;
+
+			var relativeX = (pointerPoint.Position.X - HorizontalOffset) / ZoomLevel;
+			var relativeY = (pointerPoint.Position.Y - VerticalOffset) / ZoomLevel;
+
+			ZoomLevel *= changeRatio;
+
+			HorizontalOffset = pointerPoint.Position.X - (relativeX * ZoomLevel);
+			VerticalOffset = pointerPoint.Position.Y - (relativeY * ZoomLevel);
+			return;
+		}
+
+		if (e.KeyModifiers.HasFlag(Windows.System.VirtualKeyModifiers.Shift))
+		{
+			var deltaX = GetPanDelta(pointerProperties);
+			TryUpdateOffsets(deltaX, 0);
+			return;
+		}
+
+		var deltaY = GetPanDelta(pointerProperties);
+		TryUpdateOffsets(0, deltaY);
+	}
+
+	private double GetZoomDelta(PointerPointProperties pointerProperties)
+	{
+		var delta = pointerProperties.MouseWheelDelta * ScaleWheelRatio;
+		return 1 + delta;
+	}
+
+	private double GetPanDelta(PointerPointProperties pointerProperties)
+	{
+		var delta = pointerProperties.MouseWheelDelta * PanWheelRatio;
+		return delta;
+	}
+
+	private void TryUpdateOffsets(double deltaX, double deltaY)
+	{
+		if ((deltaX > 0 && CanScrollLeft()) ||
+			(deltaX < 0 && CanScrollRight()))
+		{
+			var offset = HorizontalOffset + deltaX;
+			var max = HorizontalMaxScroll * ZoomLevel;
+			HorizontalOffset = Math.Clamp(offset, 0, max);
+		}
+
+		if ((deltaY > 0 && CanScrollUp()) ||
+			(deltaY < 0 && CanScrollDown()))
+		{
+			var offset = VerticalOffset + deltaY;
+			var max = VerticalMaxScroll * ZoomLevel;
+			VerticalOffset = Math.Clamp(offset, 0, max);
+		}
+	}
+
+	// Public methods
+	public void Initialize()
+	{
+		if (_presenter?.Content is FrameworkElement { } fe)
+		{
+			fe.UpdateLayout();
+		}
+
+		ResetZoom();
+		ResetOffset();
+		CenterContent();
+	}
+
+	public void ResetZoom() => ZoomLevel = 1;
+
+	public void ResetOffset()
+	{
+		HorizontalOffset = AdditionalMargin.Left;
+		VerticalOffset = AdditionalMargin.Top;
+	}
+
+	public void RemoveOffset()
+	{
+		HorizontalOffset = 0;
+		VerticalOffset = 0;
+	}
+
+	public void CenterContent()
+	{
+		if (IsActive && _presenter?.Content is FrameworkElement { } content)
+		{
+			HorizontalOffset = ((AvailableSize.Width - (content.ActualWidth * ZoomLevel)) / 2) + AdditionalMargin.Left;
+			VerticalOffset = ((AvailableSize.Height - (content.ActualHeight * ZoomLevel)) / 2) + AdditionalMargin.Top;
+		}
+	}
+
+	public void FitToCanvas()
+	{
+		if (IsActive)
+		{
+			var vZoom = (ActualHeight - AdditionalMargin.Top - AdditionalMargin.Bottom) / ViewPortHeight;
+			var hZoom = (ActualWidth - AdditionalMargin.Left - AdditionalMargin.Right) / ViewPortWidth;
+			var zoomLevel = Math.Min(vZoom, hZoom);
+			ZoomLevel = Math.Clamp(zoomLevel, MinZoomLevel, MaxZoomLevel);
+			CenterContent();
+		}
+	}
+
+	// Record for BoundsVisibilityData
+	public record BoundsVisibilityData(bool LeftEdge, bool TopEdge, bool RightEdge, bool BottomEdge)
```

Comment:
I'd make this a struct with one integer, and do bit operations to get/set the properties